crates/wasi-common/src/virtfs/pipe.rs

```diff
@@ -0,0 +1,412 @@
+            rights: self.rights.clone(),
+            reader: self.reader.clone(),
+        }))
+    }
+
+    fn get_file_type(&self) -> types::Filetype {
+        types::Filetype::Unknown
+    }
+
+    fn get_rights(&self) -> HandleRights {
+        self.rights.get()
+    }
+
+    fn set_rights(&self, rights: HandleRights) {
+        self.rights.set(rights)
+    }
+
+    fn advise(
+        &self,
+        _advice: types::Advice,
+        _offset: types::Filesize,
+        _len: types::Filesize,
+    ) -> Result<()> {
+        Err(Errno::Spipe)
+    }
+
+    fn allocate(&self, _offset: types::Filesize, _len: types::Filesize) -> Result<()> {
+        Err(Errno::Spipe)
+    }
+
+    fn fdstat_set_flags(&self, _fdflags: types::Fdflags) -> Result<()> {
+        // do nothing for now
+        Ok(())
+    }
+
+    fn filestat_get(&self) -> Result<types::Filestat> {
+        let stat = types::Filestat {
+            dev: 0,
+            ino: 0,
+            nlink: 0,
+            size: 0,
+            atim: 0,
+            ctim: 0,
+            mtim: 0,
+            filetype: self.get_file_type(),
+        };
+        Ok(stat)
+    }
+
+    fn filestat_set_size(&self, _st_size: types::Filesize) -> Result<()> {
+        Err(Errno::Spipe)
+    }
+
+    fn preadv(&self, buf: &mut [io::IoSliceMut], offset: types::Filesize) -> Result<usize> {
+        if offset != 0 {
+            return Err(Errno::Spipe);
+        }
+        Ok(self.reader.borrow_mut().read_vectored(buf)?)
+    }
+
+    fn seek(&self, _offset: io::SeekFrom) -> Result<types::Filesize> {
+        Err(Errno::Spipe)
+    }
+
+    fn read_vectored(&self, iovs: &mut [io::IoSliceMut]) -> Result<usize> {
+        trace!("read_vectored(iovs={:?})", iovs);
+        Ok(self.reader.borrow_mut().read_vectored(iovs)?)
+    }
+
+    fn create_directory(&self, _path: &str) -> Result<()> {
+        Err(Errno::Notdir)
+    }
+
+    fn openat(
+        &self,
+        _path: &str,
+        _read: bool,
+        _write: bool,
+        _oflags: types::Oflags,
+        _fd_flags: types::Fdflags,
+    ) -> Result<Box<dyn Handle>> {
+        Err(Errno::Notdir)
+    }
+
+    fn link(
+        &self,
+        _old_path: &str,
+        _new_handle: Box<dyn Handle>,
+        _new_path: &str,
+        _follow: bool,
+    ) -> Result<()> {
+        Err(Errno::Notdir)
+    }
+
+    fn readlink(&self, _path: &str, _buf: &mut [u8]) -> Result<usize> {
+        Err(Errno::Notdir)
+    }
+
+    fn readlinkat(&self, _path: &str) -> Result<String> {
+        Err(Errno::Notdir)
+    }
+
+    fn rename(&self, _old_path: &str, _new_handle: Box<dyn Handle>, _new_path: &str) -> Result<()> {
+        Err(Errno::Notdir)
+    }
+
+    fn remove_directory(&self, _path: &str) -> Result<()> {
+        Err(Errno::Notdir)
+    }
+
+    fn symlink(&self, _old_path: &str, _new_path: &str) -> Result<()> {
+        Err(Errno::Notdir)
+    }
+
+    fn unlink_file(&self, _path: &str) -> Result<()> {
+        Err(Errno::Notdir)
+    }
```

Comment:
Along the lines above, splitting up file descriptors into different types should also reduce the one-big-trait factor here, which should clean a lot of this up.

Reply:
Yeah, I was actually thinking about breaking up `Handle`, but since this part of the code is new to me I didn't want to stir things up too much. Are you alright with saving this for a future PR?

Reply:
Yeah, I should have clarified; I'm talking about cleanups we can do in the future here.